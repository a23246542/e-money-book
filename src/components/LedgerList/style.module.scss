.category-icon {
  width: 30px;
  text-align: center;
}
