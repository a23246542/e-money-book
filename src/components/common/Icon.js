import React from 'react';
import PropTypes from 'prop-types';
import ionicons from '../../plugin/ionicons';

// %%% props會undefined
// const Icon = ({icon, fontSize, color}) => {
const Icon = ({icon,...props}) => {
  let aa ='IosPlane';
  console.log(icon);
  
  const IconComponent = ionicons[icon];// %%icon要是字串讀取物件屬性
  return (
    <IconComponent
      {...props}
    />
  )
}

Icon.propTypes = {

}

export default Icon
