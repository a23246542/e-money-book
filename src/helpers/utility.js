export const padLeft = (number) => {
  return number < 10 ? '0' + number : String(number);
};

export const parseToYearsAndMonth = (str) => {
  const date = str ? new Date(str) : new Date();
  return {
    year: date.getFullYear(),
    month: date.getMonth() + 1,
  };
};

export const makeArrByRange = (size, startAt) => {
  let arr = [];
  for (let i = 0; i < size; i++) {
    arr[i] = startAt + i;
  }
  return arr;
};

export const flattenArr = (arr) => {
  return arr.reduce((map, item) => {
    map[item.id] = item;
    return map;
  }, {});
};

export const makeID = () => {
  // Math.random should be unique because of its seeding algorithm.
  // Convert it to base 36 (numbers + letters), and grab the first 9 characters
  // after the decimal.
  return '_' + Math.random().toString(36).substr(2, 9);
};

export const Color = {
  blue: '#347eff',
  deepBlue: '#61dafb',
  green: '#28a745',
  red: '#dc3545',
  gray: '#555',
  lightGray: '#efefef',
  white: '#fff',
};

export const PieColor = {
  // orange: '#FF8863',
  // yellow: '#FFF961',
  // green: '#4BAD6B',
  // blue: '#528ECB',
  // purple: '#9F61C2'
  orange: '#F5A536',
  yellow: '#F7D674',
  green: '#4BAD6B',
  blue: '#22A2F2',
  purple: '#57D6F2',
};
