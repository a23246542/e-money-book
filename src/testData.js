import { TYPE_OUTCOME, TYPE_INCOME } from './constants';

export const testTabs = [
  {
    text: '支出',
    value: TYPE_OUTCOME,
    iconName: 'IosCard',
  },
  {
    text: '收入',
    value: TYPE_INCOME,
    iconName: 'IosCash',
  },
  // {
  //   text:'其他',
  //   value:'other',
  //   iconName:'IosCash'
  // },
];

export const testCategories = [
  {
    name: '旅行',
    iconName: 'IosPlane',
    id: '1',
    type: 'outcome',
  },
  {
    name: '餐饮',
    iconName: 'IosRestaurant',
    id: '2',
    type: 'outcome',
  },
  {
    name: '购物',
    iconName: 'IosBasket',
    id: '3',
    type: 'outcome',
  },
  {
    name: '数码',
    iconName: 'IosPhonePortrait',
    id: '4',
    type: 'outcome',
  },
  {
    name: '工资',
    iconName: 'IosCard',
    id: '10',
    type: 'income',
  },
  {
    name: '兼职',
    iconName: 'IosCash',
    id: '11',
    type: 'income',
  },
  {
    name: '理财',
    iconName: 'LogoYen',
    id: '12',
    type: 'income',
  },
];

export const testItems = [
  {
    title: '工资',
    amount: 2000,
    date: '2020-11-18',
    monthCategory: '2020-11',
    id: '_bd16bjeen',
    cid: '10',
    timestamp: 1534550400000,
  },
  {
    title: '和哥们一起喝酒',
    amount: 300,
    date: '2020-12-20',
    monthCategory: '2020-12',
    id: '_jjfice21k',
    cid: '3',
    timestamp: 1534723200000,
  },
  {
    title: '抽奖',
    amount: 1000,
    date: '2020-12-11',
    monthCategory: '2020-12',
    id: '_1fg1wme63',
    cid: '11',
    timestamp: 1533945600000,
  },
  {
    title: '理财收入',
    amount: 300,
    date: '2020-11-15',
    monthCategory: '2020-11',
    id: '_qryggm5y8',
    cid: '12',
    timestamp: 1534291200000,
  },
  {
    title: '请别人吃饭',
    amount: 300,
    date: '2020-12-15',
    monthCategory: '2020-12',
    id: '_qryggm511',
    cid: '3',
    timestamp: 1534291200000,
  },
];
export const testItemsCopy = [
  {
    title: '工资',
    amount: 2000,
    date: '2020-11-18',
    monthCategory: '2020-11',
    id: '_bd16bjeen',
    cid: '10',
    timestamp: 1534550400000,
  },
  {
    title: '喝酒',
    amount: 300,
    date: '2020-12-20',
    monthCategory: '2020-12',
    id: '_jjfice21k',
    cid: '3',
    timestamp: 1534723200000,
  },
  {
    title: '抽奖',
    amount: 1000,
    date: '2020-12-11',
    monthCategory: '2020-12',
    id: '_1fg1wme63',
    cid: '11',
    timestamp: 1533945600000,
  },
  {
    title: '理财收入',
    amount: 300,
    date: '2020-11-15',
    monthCategory: '2020-11',
    id: '_qryggm5y8',
    cid: '12',
    timestamp: 1534291200000,
  },
  {
    title: '请别人吃饭',
    amount: 300,
    date: '2020-12-15',
    monthCategory: '2020-12',
    id: '_qryggm511',
    cid: '3',
    timestamp: 1534291200000,
  },
];
