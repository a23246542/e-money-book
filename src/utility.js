
export const padLeft = ( number ) => {
  return number < 10 ? '0'+ number : String(number)
}

export const parseToYearsAndMonth = (str) => {
  const date = str ? new Date(str) : new Date();
  return {
    year: date.getFullYear(),
    month: date.getMonth() + 1
  }
}

export const makeArrByRange = (size, startAt) => {
  let arr = [];
  for(let i = 0; i<size; i++) {
    arr[i] = startAt + i;
  }
  return arr;
}