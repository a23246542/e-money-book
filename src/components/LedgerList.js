import { useContext } from 'react';
import PropTypes from 'prop-types';
import Ionicons from '../plugin/ionicons';
import AppContext from '../AppContext';
import Icon from '../components/common/Icon';

const LedgerList = ({ items, onModifyItem, onDeleteItem }) => {
  const { IosCreate, IosCloseCircle } = Ionicons;
  // const { categories } = useContext(AppContext);//%%%items就有了
  return (
    <ul className="list-group">
      {console.log('渲染list')
      }
      {items.map((item) => {
        // const {iconName:CategoryIcon} = item.category;//%%@@不行
        // const CategoryIcon = Ionicons[item.category.iconName]
        // const CategoryIcon = Ionicons[categories[item.cid].iconName]
        // const iconName = categories[item.cid].iconName;
        return (
          <li
            className="ledger-item list-group-item d-flex align-items-center"
            key={item.id}
            data-testid={`ledger-item-${item.id}`}
          >
            <span className="col-1 badge badge-primary">
              {/* <CategoryIcon
                  iconTitle={item.category.iconName}
                  color="#fff"
                /> */}
              {<Icon icon={item.category.iconName} color="#fff" />}
            </span>
            {/* <span className="col-11 bg-primary"></span> */}
            <span className="col-5 ledger-title">{item.title}</span>
            <span className="col-2 font-weight-bold">
              {/* {item.category.type === 'income'?
                  `${item.price}元`:
                  `-${item.price}元`
                } */}
              {item.category.type === 'income' ? '+' : '-'}
              {item.amount}元
            </span>
            <span className="col-2">{item.date}</span>
            <button
              className="col-1 btn d-flex justify-content-center align-items-center"
              // style={{backgroundColor:'#28a745'}}
              // onClick={onModifyItem(item)} //會傳直接執行的函式%%%
              onClick={(e) => {
                e.preventDefault();
                onModifyItem(item);
              }}
              data-testid={`editBtn-${item.id}`}
            >
              <IosCreate
                // icon="ios-create"
                fontSize="35px"
                color="#28a745"
                style={{
                  marginRight: '5px',
                }}
              />
              編輯
            </button>
            <button
              className="col-1 btn d-flex justify-content-center align-items-center"
              onClick={(e) => {
                e.preventDefault();
                onDeleteItem(item);
              }}
            >
              <IosCloseCircle
                // icon="ios-close-circle"
                fontSize="30px"
                color="#dc3545"
                style={{
                  marginRight: '5px',
                }}
              />
              刪除
            </button>
          </li>
        );
      })}
    </ul>
  );
};

LedgerList.propTypes = {
  items: PropTypes.array.isRequired,
  onModifyItem: PropTypes.func.isRequired,
  onDeleteItem: PropTypes.func.isRequired,
};

// PriceList.defaultTypes = {

// }

export default LedgerList;
