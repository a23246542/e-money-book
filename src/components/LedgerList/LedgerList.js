import { memo } from 'react';
import PropTypes from 'prop-types';
import { IconItem } from '@/components/common';
import style from './style.module.scss';

const LedgerListComponent = ({ items, onModifyItem, onDeleteItem }) => {
  return (
    <ul className="list-group list-group-flush" data-testid="ledgerList">
      {items.map((item) => {
        return (
          <li
            className="ledger-item list-group-item d-flex align-items-center pr-0 pl-2 no-gutters border-top border-bottom"
            key={item.id}
            data-testid={`ledger-item-${item.id}`}
            title="ledger-item"
          >
            <div className="col-1 text-center">
              <span
                className={`${style['category-icon']} badge badge-primary rounded-circle`}
              >
                <IconItem
                  icon={item.category.iconName}
                  color="#fff"
                  fontSize="20px"
                />
              </span>
            </div>
            <div className="col-5 ledger-title">{item.title}</div>
            <div className="col-2 font-weight-bold">
              {item.category.type === 'income' ? '+' : '-'}
              {item.amount}
            </div>
            <span className="col-2">{item.date.replace('2021-', '')}</span>
            <button
              className="col-1 btn btn-edit d-flex justify-content-center align-items-center"
              onClick={(e) => {
                e.preventDefault();
                onModifyItem(item);
              }}
              data-test="editBtn"
              data-testid={`editBtn-${item.id}`}
              title="editBtn"
            >
              <IconItem
                icon="IosCreate"
                fontSize="23px"
                color="#28a745"
                style={{
                  marginRight: '5px',
                }}
              />
              {/* 編輯 */}
            </button>
            <button
              className="col-1 btn btn-delete d-flex justify-content-center align-items-center"
              onClick={(e) => {
                e.preventDefault();
                onDeleteItem(item);
              }}
              data-test="deleteBtn"
            >
              <IconItem
                icon="IosCloseCircle"
                fontSize="20px"
                color="#dc3545"
                style={{
                  marginRight: '5px',
                }}
              />
              {/* 刪除 */}
            </button>
          </li>
        );
      })}
    </ul>
  );
};

LedgerListComponent.propTypes = {
  items: PropTypes.array.isRequired,
  onModifyItem: PropTypes.func.isRequired,
  onDeleteItem: PropTypes.func.isRequired,
};

export const LedgerList = memo(LedgerListComponent);
