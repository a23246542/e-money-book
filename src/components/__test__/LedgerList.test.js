import React from 'react';
import Ionicons from '../../plugin/ionicons';
import IosPlane from 'react-ionicons/lib/IosPlane';
import { flattenArr } from '../../utility';
import { shallow, mount } from 'enzyme';
import LedgerList from '../LedgerList';
// import AppContext, { Provider } from '../../AppContext';
import { testCategories, testItems } from '../../helpers/testData';
import Icon from '../../components/common/Icon';

const category = {
  1: {
    name: '旅行',
    type: 'outcome',
    iconName: 'IosPlane',
  },
  2: {
    name: '領薪水',
    type: 'income',
    iconName: 'IosPlane',
  },
  3: {
    name: '投資',
    type: 'income',
    iconName: 'IosPlane',
  },
};

const items = [
  {
    id: 1,
    title: '去雲南旅遊',
    price: 200,
    date: '2020-11-26',
    categoryId: 1,
  },
  {
    id: 2,
    title: '領薪水',
    price: 1000,
    date: '2020-11-26',
    categoryId: 2,
  },
  {
    id: 3,
    title: '去台灣旅遊',
    price: 400,
    date: '2020-11-27',
    categoryId: 1,
  },
];

const listWithCategory = testItems.map((item) => {
  return {
    ...item,
    category: flattenArr(testCategories)[item.cid],
  };
});

const props = {
  items: listWithCategory,
  onModifyItem: jest.fn(),
  onDeleteItem: jest.fn(),
};

let wrapper;

describe('test LedgerList component', () => {
  beforeEach(() => {
    wrapper = mount(<LedgerList {...props} />);
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render component to match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correct ledger items length', () => {
    expect(wrapper.find('.list-group-item').length).toEqual(
      listWithCategory.length
    );
  });

  it('should render correct icon and ledger for each item', () => {
    const iconList = wrapper.find('.list-group-item').first().find(Icon);
    expect(iconList.first().props().icon).toEqual(
      listWithCategory[0].category.iconName
    );
  });

  it('should trigger the correct function callbacks', () => {
    const firstItem = wrapper.find('.ledger-item').first();
    firstItem
      .find('[data-test="editBtn"]')
      .first()
      .simulate('click', { preventDefault: () => {} });
    expect(props.onModifyItem).toHaveBeenCalledWith(listWithCategory[0]);
    firstItem
      .find('[data-test="deleteBtn"]')
      .first()
      .simulate('click', { preventDefault: () => {} });
    expect(props.onDeleteItem).toHaveBeenCalledWith(listWithCategory[0]);
  });
});
