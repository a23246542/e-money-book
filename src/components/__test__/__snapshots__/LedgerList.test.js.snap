// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test LedgerList component should render component to match snapshot 1`] = `ShallowWrapper {}`;
